File: Video_ZeromqSubscriber.cpp
#include <iostream>
#include <opencv2/opencv.hpp>
#include <zmq.hpp>

int main() 
{
  // Initialize ZeroMQ context
  zmq::context_t context(1);

  // Create a ZeroMQ subscriber socket
  zmq::socket_t subscriber(context, ZMQ_SUB);
  subscriber.connect("tcp://localhost:5555");
  subscriber.setsockopt(ZMQ_SUBSCRIBE, "", 0);

  // Create an OpenCV window
  cv::namedWindow("Video Frame", cv::WINDOW_AUTOSIZE);

  while (true) 
  {
    // Receive video frame from ZeroMQ
    zmq::message_t frame;
    subscriber.recv( & frame);

    // Convert the received data to an OpenCV Mat
    cv::Mat image = cv::imdecode(cv::Mat(1, frame.size(), CV_8UC1, frame.data()), cv::IMREAD_COLOR);

    // Display the video frame
    cv::imshow("Video Frame", image);

    // Wait for a key press to exit the loop
    if (cv::waitKey(1) == 27) 
    {
      break;
    }
  }

  return 0;
}

File: Video_ZeromqSubscriber.py

import cv2
import zmq

# Create a ZeroMQ context
context = zmq.Context()

# Create a ZeroMQ socket for receiving video frames
socket = context.socket(zmq.SUB)
socket.connect("tcp://localhost:5555")  # Replace with the appropriate address and port
socket.setsockopt(zmq.SUBSCRIBE, b'')  # Subscribe to all messages

while True:
    # Receive the video frame from the socket
    frame_data = socket.recv()

    # Decode the received data into a NumPy array
    frame = cv2.imdecode(np.frombuffer(frame_data, dtype=np.uint8), cv2.IMREAD_COLOR)

    # Display the video frame
    cv2.imshow('Video', frame)

    # Check for the 'q' key to exit
    if cv2.waitKey(1) & 0xFF == ord('q'):
        break

# Clean up
cv2.destroyAllWindows()
socket.close()
context.term()
